hail/src/main/scala/is/hail/lir/X.scala

```diff
@@ -466,93 +441,135 @@ abstract class StmtX extends X {
     next = null
     prev = null
   }
+
+  def replace(x: StmtX): Unit = {
+    assert(x.parent == null)
+    assert(parent != null)
+
+    x.next = next
+    x.prev = prev
+    x.parent = parent
+
+    if (parent.first == this)
+      parent.first = x
+    if (parent.last == this)
+      parent.last = x
+    if (next != null)
+      next.prev = x
+    if (prev != null)
+      prev.next = x
+
+    next = null
+    prev = null
+    parent = null
+  }
+
+  def insertBefore(x: StmtX): Unit = {
+    assert(parent != null)
+    assert(x.parent == null)
+
+    x.next = this
+    x.parent = parent
+    if (prev != null) {
+      x.prev = prev
+      prev.next = x
+    } else {
+      parent.first = x
+    }
+    prev = x
+  }
 }
 
-abstract class ControlX extends StmtX
+abstract class ControlX extends StmtX {
+  def targetArity(): Int
+
+  def target(i: Int): Block
+
+  def setTarget(i: Int, b: Block): Unit
+}
 
 abstract class ValueX extends X {
   var parent: X = _
 
   def ti: TypeInfo[_]
-}
-
-class GotoX extends ControlX {
-  var _L: Block = _
 
-  def L: Block = _L
+  def remove(): Unit = {
+    var i = 0
+    while (parent.children(i) ne this)
+      i += 1
+    parent.setChild(i, null)
+    assert(parent == null)
```

Comment:
This can't be true, can it?

Reply:
I'm still confused by this. We just called a method on `parent`, how can `parent` be null?

Reply:
`this` is the `i`th child.  `setParent` removes the existing child, which means setting `parent = null`, because the node is no longer a child of the parent.

Reply:
Sorry, I see it now.